### ISSUE TITLE
Gradient Relationship between "aux_estimate.grad", "mean.grad" and "perturbation.grad"?

### ISSUE THREAD

**Issue by HenryPengZou** [eth-sri/probabilistic-forecasts-attacks#6]
https://github.com/eth-sri/probabilistic-forecasts-attacks/blob/df78ffd7375089bb7b38fbf55bb840b7699365eb/Finance/attack/attacks.py#L91-L116

Hi, one more thing. 

I suppose the first loss.backward() here calculates ''mean.grad'' , aux_estimate.backward() calculates "aux_estimate.grad", and the second loss.backward() computes 'perturbation.grad'. Could you provide the relationship between these three gradient?

My guess is : perturbation.grad = mean.grad * aux_estimate.grad, is this correct?
i.e.: 
![CodeCogsEqn](https://user-images.githubusercontent.com/79878129/126737998-4ad07537-93d4-4528-99e2-3b5d8c7b9e3c.png)

**Comment by rdang-nhu** [comment]
Hi. This is related to issues #4 and #5. The loss function to differentiate is
<img width="385" alt="Screenshot 2021-07-23 at 17 34 23" src="https://user-images.githubusercontent.com/18661420/126805991-c872c8a4-4139-4bfe-8ec4-29c05fcb1da0.png">

1. aux_estimate.backward() computes the gradient _mean (the expectation) -> perturbation (delta)_  **using the score function estimator** and places it in `pertubation.grad` 
2. The first loss.backward() computes the gradient _loss (obj) -> mean (the expectation)_ and places it in `mean.grad`, which is then multiplied to `pertubation.grad` to apply the chain rule

Together, 1. and 2. compute the gradient _loss (obj) -> perturbation (delta)_, **through the function phi that measures distance to the target t**

3. The second loss.backward() computes the gradient  _loss (obj) -> perturbation (delta)_ , and adds it to`pertubation.grad` but only **through the norm**. (when calling `.backward()` several times, the different gradients are accumulated by summation).

This convoluted way of computing the gradient is needed because with the score function estimator, we can not rely on autograd to compute 1. automatically.

**Comment by HenryPengZou** [comment]
Great explanation, very clear! At first, I was confused by 'aux_estimate' computes 'mean.grad'. But my doubt and questions are solved by your most answer, very clear, like it. I appreciate your effort in editing this answer times and times to make it become more clear and perfect. Thanks again!🌷
